diff --git a/Cargo.lock b/Cargo.lock
index 059b203f..aa3d9de6 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -5014,12 +5014,21 @@ dependencies = [
  "ignore",
  "itertools",
  "regex",
+ "rustc_ast",
+ "rustc_ast_pretty",
+ "rustc_data_structures",
+ "rustc_errors",
+ "rustc_expand",
+ "rustc_parse",
+ "rustc_session",
+ "rustc_span",
  "rustfmt-config_proc_macro",
  "semver",
  "serde",
  "serde_json",
  "tempfile",
  "term",
+ "thin-vec",
  "thiserror 1.0.69",
  "toml 0.9.8",
  "tracing",
diff --git a/src/tools/rustfmt/Cargo.toml b/src/tools/rustfmt/Cargo.toml
index 6ac140f2..e60a4e26 100644
--- a/src/tools/rustfmt/Cargo.toml
+++ b/src/tools/rustfmt/Cargo.toml
@@ -62,6 +62,18 @@ unicode-properties = { version = "0.1", default-features = false, features = ["g
 rustfmt-config_proc_macro = { version = "0.3", path = "config_proc_macro" }
 semver = "1.0.21"
 
+rustc_ast = { path = "../../../compiler/rustc_ast" }
+rustc_ast_pretty = { path = "../../../compiler/rustc_ast_pretty" }
+rustc_data_structures = { path = "../../../compiler/rustc_data_structures" }
+rustc_errors = { path = "../../../compiler/rustc_errors" }
+rustc_expand = { path = "../../../compiler/rustc_expand" }
+rustc_parse = { path = "../../../compiler/rustc_parse" }
+rustc_session = { path = "../../../compiler/rustc_session" }
+rustc_span = { path = "../../../compiler/rustc_span" }
+
+# Use the version from compiler/rustc_ast/Cargo.toml.
+thin-vec = "0.2.15"
+
 [dev-dependencies]
 tempfile = "3.23.0"
 
diff --git a/src/tools/rustfmt/src/bin/main.rs b/src/tools/rustfmt/src/bin/main.rs
index 6c48301f..b2ff011b 100644
--- a/src/tools/rustfmt/src/bin/main.rs
+++ b/src/tools/rustfmt/src/bin/main.rs
@@ -1,5 +1,3 @@
-#![feature(rustc_private)]
-
 use anyhow::{Result, format_err};
 
 use io::Error as IoError;
@@ -22,14 +20,9 @@
     FormatReportFormatterBuilder, Input, Session, StyleEdition, Verbosity, Version, load_config,
 };
 
-const BUG_REPORT_URL: &str = "https://github.com/rust-lang/rustfmt/issues/new?labels=bug";
 
-// N.B. these crates are loaded from the sysroot, so they need extern crate.
-extern crate rustc_driver;
 
 fn main() {
-    rustc_driver::install_ice_hook(BUG_REPORT_URL, |_| ());
-
     tracing_subscriber::fmt()
         .with_env_filter(EnvFilter::from_env("RUSTFMT_LOG"))
         .init();
diff --git a/src/tools/rustfmt/src/lib.rs b/src/tools/rustfmt/src/lib.rs
index 942b42ec..7fe2146c 100644
--- a/src/tools/rustfmt/src/lib.rs
+++ b/src/tools/rustfmt/src/lib.rs
@@ -5,21 +5,6 @@
 #![allow(clippy::match_like_matches_macro)]
 #![allow(unreachable_pub)]
 
-// N.B. these crates are loaded from the sysroot, so they need extern crate.
-extern crate rustc_ast;
-extern crate rustc_ast_pretty;
-extern crate rustc_data_structures;
-extern crate rustc_errors;
-extern crate rustc_expand;
-extern crate rustc_parse;
-extern crate rustc_session;
-extern crate rustc_span;
-extern crate thin_vec;
-
-// Necessary to pull in object code as the rest of the rustc crates are shipped only as rmeta
-// files.
-#[allow(unused_extern_crates)]
-extern crate rustc_driver;
 
 use std::cell::RefCell;
 use std::cmp::min;
